File: workers/download-tracker/src/update.js
/**
 * Client update prompt for GodLock. Calls runtime
 * GET /v1/update/check?slug=godlock&version=… when online.
 * Surfaces counted /download when update_available. Never overwrites.
 * Author: Aziel Eliab.
 */

export const RUNTIME_HOST = "https://aziel-runtime.vibelock.workers.dev";
export const UPDATE_CHECK_PATH = "/v1/update/check";
export const PULL_PATH = "/v1/pull/godlock";
export const COUNTED_DOWNLOAD = "https://godlock-download-tracker.vibelock.workers.dev/download";
export const SLUG = "godlock";
export const UA = { "User-Agent": "Mozilla/5.0", Accept: "application/json" };

export function compareVersions(left, right) {
  const parts = (raw) => String(raw || "")
    .trim()
    .replace(/^[vV]/, "")
    .split(".")
    .map((chunk) => {
      const m = String(chunk).match(/^\d+/);
      return m ? parseInt(m[0], 10) : null;
    })
    .filter((n) => n != null);
  const a = parts(left);
  const b = parts(right);
  if (!a.length || !b.length) return 0;
  const n = Math.max(a.length, b.length);
  while (a.length < n) a.push(0);
  while (b.length < n) b.push(0);
  for (let i = 0; i < n; i++) {
    if (a[i] > b[i]) return 1;
    if (a[i] < b[i]) return -1;
  }
  return 0;
}

export function parseUpdateDoc(body, current) {
  if (!body || typeof body !== "object") return null;
  const latest = String(body.latest || body.latest_version || body.version || "").trim();
  const installed = String(body.current || body.installed || current || "").trim();
  const download = String(body.download || body.download_url || COUNTED_DOWNLOAD).trim() || COUNTED_DOWNLOAD;
  let flagged = body.update_available === true || String(body.update_available || "").toLowerCase() === "true";
  if (!flagged && latest && installed && compareVersions(latest, installed) > 0) flagged = true;
  if (body.error && !latest) return null;
  if (!latest && !flagged) return null;
  return {
    ok: true,
    slug: SLUG,
    version: installed || current,
    latest: latest || installed || current,
    update_available: !!flagged,
    download,
    forced: false,
    author: "Aziel Eliab",
  };
}

export function formatUpdatePrompt(doc) {
  if (!doc || !doc.update_available) return "";
  return (
    "Update available: GodLock " + (doc.latest || "") +
    " (you have " + (doc.version || "") + "). " +
    "Counted download (no silent overwrite): " + (doc.download || COUNTED_DOWNLOAD)
  );
}

async function readJson(res) {
  if (!res || !res.ok) return null;
  return res.json().catch(() => null);
}

export function updateCheckUrl(version) {
  return RUNTIME_HOST + UPDATE_CHECK_PATH + "?slug=" + encodeURIComponent(SLUG) + "&version=" + encodeURIComponent(version);
}

export async function checkGodlockUpdate(deps = {}) {
  const version = String(deps.version || "0.1.0");
  const httpFetch = deps.fetch || globalThis.fetch;
  const timeoutMs = deps.timeoutMs != null ? deps.timeoutMs : 2500;
  if (typeof httpFetch !== "function") return null;

  const urls = [updateCheckUrl(version), RUNTIME_HOST + PULL_PATH];
  for (const url of urls) {
    try {
      const ac = typeof AbortController === "function" ? new AbortController() : null;
      const timer = ac ? setTimeout(() => ac.abort(), timeoutMs) : null;
      const init = { headers: UA };
      if (ac) init.signal = ac.signal;
      const res = await httpFetch(url, init);
      if (timer) clearTimeout(timer);
      const parsed = parseUpdateDoc(await readJson(res), version);
      if (parsed) {
        parsed.source = url.includes(UPDATE_CHECK_PATH) ? "update-check" : "pull";
        parsed.forced = false;
        parsed.prompt = formatUpdatePrompt(parsed);
        return parsed;
      }
    } catch { /* try next / stay silent when offline */ }
  }
  return null;
}

File: godlock/update.py
"""Optional online update prompt for GodLock.

Calls aziel-runtime ``GET /v1/update/check?slug=godlock&version=…`` when a
network is available. Falls back to ``GET /v1/pull/godlock``. Surfaces a
counted ``/download`` link when ``update_available``. Never overwrites the
local install. Author: Aziel Eliab.
"""

from __future__ import annotations

import json
import urllib.error
import urllib.parse
import urllib.request
from typing import Any, Callable

from godlock import __version__

RUNTIME_HOST = "https://aziel-runtime.vibelock.workers.dev"
UPDATE_CHECK_PATH = "/v1/update/check"
PULL_PATH = "/v1/pull/godlock"
DOWNLOAD = "https://godlock-download-tracker.vibelock.workers.dev/download"
SLUG = "godlock"
USER_AGENT = "Mozilla/5.0"
DEFAULT_TIMEOUT_S = 2.5

Fetcher = Callable[[str], dict[str, Any] | None]


def compare_versions(left: str, right: str) -> int:
    """Return 1 if left > right, -1 if left < right, 0 if equal / unparsable."""
    def parts(raw: str) -> list[int]:
        out: list[int] = []
        for chunk in str(raw or "").strip().lstrip("vV").split("."):
            digits = ""
            for ch in chunk:
                if ch.isdigit():
                    digits += ch
                else:
                    break
            if digits == "":
                break
            out.append(int(digits))
        return out

    a, b = parts(left), parts(right)
    if not a or not b:
        return 0
    n = max(len(a), len(b))
    a.extend([0] * (n - len(a)))
    b.extend([0] * (n - len(b)))
    if a > b:
        return 1
    if a < b:
        return -1
    return 0


def _truthy(value: Any) -> bool:
    if value is True:
        return True
    if isinstance(value, str) and value.strip().lower() in {"1", "true", "yes"}:
        return True
    return False


def parse_update_doc(body: Any, current: str) -> dict[str, Any] | None:
    if not isinstance(body, dict):
        return None
    latest = str(body.get("latest") or body.get("latest_version") or body.get("version") or "").strip()
    installed = str(body.get("current") or body.get("installed") or current or "").strip()
    download = str(body.get("download") or body.get("download_url") or DOWNLOAD).strip() or DOWNLOAD
    flagged = _truthy(body.get("update_available"))
    if not flagged and latest and installed and compare_versions(latest, installed) > 0:
        flagged = True
    if body.get("error") and not latest:
        return None
    if not latest and not flagged:
        return None
    return {
        "ok": True,
        "slug": SLUG,
        "version": installed or current,
        "latest": latest or installed or current,
        "update_available": bool(flagged),
        "download": download,
        "forced": False,
        "author": "Aziel Eliab",
    }


def _default_fetch(url: str, timeout: float = DEFAULT_TIMEOUT_S) -> dict[str, Any] | None:
    req = urllib.request.Request(
        url,
        headers={"User-Agent": USER_AGENT, "Accept": "application/json"},
        method="GET",
    )
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            if getattr(resp, "status", 200) >= 400:
                return None
            raw = resp.read().decode("utf-8")
    except (urllib.error.URLError, TimeoutError, OSError, ValueError):
        return None
    try:
        data = json.loads(raw)
    except json.JSONDecodeError:
        return None
    return data if isinstance(data, dict) else None


def check_update(
    version: str | None = None,
    *,
    fetch: Fetcher | None = None,
    timeout: float = DEFAULT_TIMEOUT_S,
) -> dict[str, Any] | None:
    """Return an update record or None when offline / unchanged / unreadable.

    Never raises. Never downloads a package. ``forced`` is always False.
    """
    current = str(version or __version__)
    getter = fetch or (lambda url: _default_fetch(url, timeout=timeout))
    check_url = (
        f"{RUNTIME_HOST}{UPDATE_CHECK_PATH}"
        f"?slug={SLUG}&version={urllib.parse.quote(current, safe='')}"
    )
    parsed = parse_update_doc(getter(check_url), current)
    if parsed is None:
        parsed = parse_update_doc(getter(f"{RUNTIME_HOST}{PULL_PATH}"), current)
    if parsed is None:
        return None
    parsed["forced"] = False
    return parsed


def format_prompt(doc: dict[str, Any] | None) -> str:
    if not doc or not doc.get("update_available"):
        return ""
    latest = doc.get("latest") or ""
    current = doc.get("version") or __version__
    download = doc.get("download") or DOWNLOAD
    return (
        f"Update available: GodLock {latest} (you have {current}). "
        f"Counted download (no silent overwrite): {download}"
    )
